src/esp/assets/ResourceManager.h

```diff
@@ -366,86 +359,99 @@ class ResourceManager {
-   * identified by the string key
+   * identified by the string key.
    * @param primTemplateHandle the string key of the attributes desired
-   * @return the desired primitive attributes
+   * @return the desired primitive attributes, or nullptr if does not exist
    */
   AbstractPrimitiveAttributes::ptr getPrimitiveTemplateAttributes(
       const std::string& primTemplateHandle) const {
+    CORRADE_ASSERT(
+        (primitiveAssetsTemplateLibrary_.count(primTemplateHandle) > 0),
+        "ResourceManager::getPrimitiveTemplateAttributes : Unknown primitive "
+        "template handle : "
+            << primTemplateHandle,
+        nullptr);
     return primitiveAssetsTemplateLibrary_.at(primTemplateHandle);
   }
 
   /**
    * @brief Get a reference to the physics object template for the asset
    * identified by the key, templateHandle.  physicsObjTemplateLibrary_
    *
-   * Can be used to manipulate an object
-   * template before instancing new objects.
+   * Can be used to manipulate an object template before instancing new objects.
    * @param templateHandle The key referencing the asset in @ref
    * physicsObjTemplateLibrary_.
-   * @return A mutable reference to the object template for the asset.
+   * @return A mutable reference to the object template for the asset, or
+   * nullptr if does not exist
    */
   PhysicsObjectAttributes::ptr getPhysicsObjectAttributes(
       const std::string& templateHandle) const {
+    CORRADE_ASSERT(
+        (physicsObjTemplateLibrary_.count(templateHandle) > 0),
+        "ResourceManager::getPhysicsObjectAttributes : Unknown template handle "
+        ": " << templateHandle,
+        nullptr);
     return physicsObjTemplateLibrary_.at(templateHandle);
   }
 
   /**
    * @brief Get a reference to the physics object template for the asset
    * identified by the objectTemplateID.
    *
-   * Can be used to manipulate an object
-   * template before instancing new objects.
-   * @param ObjTmplID The key referencing the asset in @ref
+   * Can be used to manipulate an object template before instancing new objects.
+   * @param objectTemplateID The key referencing the asset in @ref
    * physicsObjTemplateLibrary_.
-   * @return A mutable reference to the object template for the asset.
+   * @return A mutable reference to the object template for the asset, or
+   * nullptr if does not exist
    */
   PhysicsObjectAttributes::ptr getPhysicsObjectAttributes(
       const int objectTemplateID) const {
-    return physicsObjTemplateLibrary_.at(
-        getObjectTemplateHandle(objectTemplateID));
+    std::string key = getObjectTemplateHandle(objectTemplateID);
+    CORRADE_ASSERT((physicsObjTemplateLibrary_.count(key) > 0),
+                   "ResourceManager::getPhysicsObjectAttributes : Unknown "
+                   "object template ID:"
+                       << objectTemplateID,
+                   nullptr);
+    return physicsObjTemplateLibrary_.at(key);
   }
```

Comment:
I would personally de-inline all the `getPhysics*()` functions into the cpp file.

Reply:
How come? I am interested in hearing your philosophy about this.   My (limited) perspective is that it is ok to inline methods that are < 10 lines, especially if they are just glorified accessors, and I am trying to remove some of the clutter from ResourceManager.cpp so I can focus on the actual functionality.  

I'd be the first to admit, though, that I am no expert, so if you have any insights to share I would really appreciate it :).

Reply:
these methods aren't too long from the visual perspective, but:

- if they use STL containers in any way (which these do), there's a lot of template onions to be parsed and instantiated every time the compiler sees the definition, slowing down compile times compared to if the method is de-inlined in the cpp
- additionally, them being defined inside the class definiton makes them implicitly inline, which inflates binary size and slows down link times (especially when LTO is enabled, which is for python bindings), as the same code is pasted into every call site

that said, this project doesn't have the best compile times anyway (and can't have, due to heavy header-only dependencies like eigen and pybind), so one change like this won't make much difference, but it adds up i'd say

Reply:
Thanks for this info.  I was unaware of the impact the stl containers would have, something to research :).